File: modules/cpp/LitUtility.cpp
#include <vector>
#include <cstdint>
#include <iostream>
#include <cmath>

/**
 * @brief りっとー便利ツール詰め合わせ
 * @details 状態を持たない関数のうち、ひとまとめにするほど規模の大きくないこまごまとした関数を置く
 */
namespace LitUtility
{
    /**
     * 二分探索関数
     * 評価関数evalがtrueになる範囲とfalseになる範囲の境界値を求める
     * @param ok trueになる片極端
     * @param ng falseになる片極端
     * @param eval 評価関数
     */
    template <class Func>
    int64_t Search(int64_t ok, int64_t ng, Func eval)
    {
        while (1 < std::abs(ok - ng))
        {
            int64_t mid = (ok + ng) / 2;
            if (eval(mid))
            {
                ok = mid;
            }
            else
            {
                ng = mid;
            }
        }
        return ok;
    }
    /**
     * 値と桁数からビット列を生成する。
     * e.g.,(5,3) -> {true,false,true}
     */
    std::vector<bool> GenerateBit(int64_t value, const int64_t digits)
    {
        std::vector<bool> result(digits);
        for (int64_t i = 0; i < digits; ++i)
        {
            result[digits - i - 1] = (value & 1);
            value >>= 1;
        }
        ++value;
        return result;
    }
    /**
     * @brief コンテナの中の最大値を求める
     * @param 求めたいコンテナ(std::vector等)
     */
    template <class T>
    typename T::value_type Max(const T &target)
    {
        typename T::value_type nowMax = std::numeric_limits<typename T::value_type>::min();
        for (auto &i : target)
        {
            if (nowMax < i)
            {
                nowMax = i;
            }
        }
        return nowMax;
    }
    /**
     * @brief コンテナの中の最小値を求める
     * @param 求めたいコンテナ(std::vector等)
     */
    template <class T>
    typename T::value_type Min(const T &target)
    {
        typename T::value_type nowMin = std::numeric_limits<typename T::value_type>::max();
        for (auto &i : target)
        {
            if (i < nowMin)
            {
                nowMin = i;
            }
        }
        return nowMin;
    }
    template <class T>
    typename T::value_type Sum(const T &target)
    {
        typename T::value_type now = 0;
        for (auto &i : target)
        {
            now += i;
        }
        return now;
    }

    /**
     * @brief 順列を計算する
     * @param n n 母数
     * @param r r 項目数
     * @return nPrの計算結果
     *
     */
    int64_t Permutation(int64_t n, int64_t r)
    {
        int64_t result = 1;
        for (int64_t i = 0; i < r; ++i)
        {
            result *= (n - i);
        }
        return result;
    }
    /**
     * @brief 階乗を計算する
     * @param n 値
     * @return n!の計算結果
     *
     */
    int64_t Factorial(int64_t n)
    {
        return Permutation(n, n);
    }
    /**
     * @brief 組み合わせを計算する
     * @param n 母数
     * @param r 選択数
     * @return nCrの計算結果
     *
     */
    int64_t Combination(int64_t n, int64_t r)
    {
        int64_t result = 1;
        for (int64_t i = 0; i < r; ++i)
        {
            result *= (n - i);
            result /= i + 1;
        }
        return result;
    }
    /**
     * @brief xよりyが大きいときのみ更新する
     * @return xよりyのほうが大きいかどうか
     */
    int64_t ChangeMax(int64_t &x, int64_t &y)
    {
        if (x < y)
        {
            x = y;
            return true;
        }
        return false;
    }
    /**
     * @brief xよりyが小さいときのみ更新する
     * @return xよりyのほうが小さいかどうか
     */
    int64_t ChangeMin(int64_t &x, int64_t &y)
    {
        if (y < x)
        {
            y = x;
            return true;
        }
        return false;
    }
    /**
     * @brief 条件がtrueのときにYesと出力する
     * @details if分岐中にYesを吐き出したい
     * @param isYes Yesを吐き出す条件
     * @return isYesの中身
     */
    bool Yes(bool isYes = true)
    {
        if (isYes)
        {
            std::cout << "Yes" << std::endl;
        }
        return isYes;
    }
    /**
     * @brief 条件がtrueのときにNoと出力する
     * @details if分岐中にNoを吐き出したい
     * @param isNo Noを吐き出す条件
     * @return isNoの中身
     */
    bool No(bool isNo = true)
    {
        if (isNo)
        {
            std::cout << "No" << std::endl;
        }
        return isNo;
    }
    /**
     * @brief 条件がtrueのときにYes,そうでないときにNoと出力する
     * @param isYes Yesを吐き出す条件
     * @return isYesの中身
     */
    bool YesNo(bool isYes)
    {
        if (!isYes)
        {
            Yes();
        }
        else
        {
            No();
        }
        return isYes;
    }
}
namespace Lit = LitUtility;

File: modules/py/LitUtility.py
import math


class LitUtility:
    def Search(ok:int,ng:int,f:bool)->int:
        """二分探索を行う関数
        単調増加の範囲においてokのうちいちばんngに近いものの値を返す
        利用例:
        -  lambda i:a[i]<x xを含まない最大のiを返す
        -  lambda i:a[i]<=x xを含む最大のiを返す

        Args:
            -  ok (int): 評価関数fに渡したときに必ずTrueを返すことが保証されている値
            -  ng (int): 評価関数fに渡したときに必ずFalseを返すことが保証されている値
            -  f (bool): 評価関数(引数1/戻り値bool)

        Returns:
            -  int: 結果
        """
        while 1<abs(ok-ng):
            mid=(ng+ok)//2
            if f(mid):
                ok=mid
            else:
                ng=mid
        return ok

    def Compression(A: list) -> list:
        """座標圧縮
        座標を左上に詰める。
        Aをコピー、重複を消し、ソートする。

        Args:
            -  A (list): 圧縮する配列が入ったリスト

        Returns:
            -  list: 圧縮結果
        """
        B = sorted(list(set(A.copy())))
        result = []  # 結果となるべき配列
        for i in A:
            result.append(LitUtility.Index(B, i))  # Aの要素がBのうち何番目にあたるかを返す
        return result


    def Index(a: list, i: int) -> int:  # 簡易二分探索
        # 二分探索で配列の中のiが何番目にあるかを返す
        # →二分探索モジュールと同様のためドキュメント不在
        f = lambda x: a[x] <= i
        ok = -1
        ng = len(a)
        while 1 < abs(ok - ng):
            mid = (ng + ok) // 2
            if f(mid):
                ok = mid
            else:
                ng = mid
        else:
            return ok
    def Factorial(n: int) -> int:
        """
        n!の結果を返す。

        より具体的には整数nの階乗の値を返す。

        Args:
            -  n(int):nの値

        Returns:
            -  int: 計算結果
        """
        return LitUtility.Permutation(n, n)

    def Permutation(n: int, k: int) -> int:
        """
        nPkの値を返す。
        より具体的にはn個の整数からk個を順番に並べられる通り数を返す。

        Args:
            -  n(int):nの値
            -  k(int):kの値

        Returns:
            -  int: 計算結果
        """
        result = 1
        for i in range(n, n - k, -1):
            result *= i
        return result

    def Combination(n: int, k: int) -> int:
        """
        nCkの値を返す。
        より具体的にはn個の整数からk個を選ぶ組み合わせ数を返す。

        Args:
            -  n(int):nの値
            -  k(int):kの値

        Returns:
            -  int: 計算結果
        """
        return LitUtility.Permutation(n, k) // LitUtility.Factorial(k)

    def CalcGCD(value_a: int, value_b: int) -> int:
        """互除法で最大公約数を求める関数

        Args:
            -  value_a (int): 求めたい値A
            -  value_b (int): 求めたい値B

        Returns:
            -  int: AとBの最大公約数
        """
        if value_a < value_b:
            value_a, value_b, value_b, value_a
        while value_b != 0:
            temp = value_b
            value_b = value_a % value_b
            value_a = temp
        return value_a

    def Prime(num:int) -> list:
        """素因数分解の結果を返す
        ※利用するためには import mathが必須

        Args:
            -  num (int): 合成数または素数

        Returns:
            -  list: 分解した素数を返す
        """
        result=[]
        while num%2==0:
            num//=2
            result.append(2)
        c=3
        while c<=math.sqrt(num):
            if num%c==0:
                num//=c
                result.append(c)
            else:
                c+=2
        if num!=1:
            result.append(num)
        return result


